### PR TITLE
Add support for incremental SMT traces containing C_bool

### PR DESCRIPTION
**Pull request by thomasspriggs**
An "unsupported type" invariant violation would previously have been
encountered when attempting to generate traces featuring `c_bool_typet`.
This is due to `c_bool_typet` not being castable to
`integer_bitvector_typet`.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by TGWDB — APPROVED** [review]
LGTM

It would have been prettier to have split commits for parts that are different (e.g. refactor, code reorder, tests), but for a PR this small it's fine.

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
   {
     if(
       const auto integer_type =
```
Comment (NlightNFotis) [inline src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: Should the variable name be changed then, if it no longer strictly corresponds to an integer type?

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdbool_example.c [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#include <stdbool.h>
+
+int main()
+{
+  int x, y;
+  bool equal = x == y;
+  __CPROVER_assert(equal, "Assert of integer equality.");
+  __CPROVER_assert(!equal, "Assert of not integer equality.");
```
Comment (NlightNFotis) [inline regression/cbmc-incr-smt2/nondeterministic-int-assert/stdbool_example.c:8]: ⛏️ What does `not integer equality` mean? Do you mean `inequality` in this context?

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp (hunk 1 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: Wait I'm reading more on this, and I'm becoming more and more confused...

It seems that `bool` maps to `_Bool`, which is a integral type with a two value domain: `{0, 1}`. If that's the case, why is it not transferable to an `integer`?

Is it because we have messed up the handling for that internally (in CBMC that is)?

File: regression/cbmc-incr-smt2/nondeterministic-int-assert/stdbool_example.c (hunk 2 above)
Comment (thomasspriggs) [reply regression/cbmc-incr-smt2/nondeterministic-int-assert/stdbool_example.c:8]: I mean the not operator is being applied to the result of testing for integer equality. This is equivalent to testing for inequality with the `!=` operator, so the distinction is not really important here. The main purpose of this assertion is to ensure that the output traces include instances where equal is assigned a true value and equal is assigned a false value.

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp (hunk 1 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: To answer your original question - Yes. I should rename this.

`_Bool` is a type which can only be set to 0 or 1 - "two value domain: {0, 1}". However you can expect that it will be stored in 1 or more bytes of memory, rather than packed into individual bits. The (unused) most significant bits will then be packed with `0`s to fill the full piece of memory allocated. For example if `_Bool` is 8 bits wide then the values will be `00000001` and `00000000` for true and false respectively. The property of allowing values 0 and 1 exclusively is satisfied by mapping all non-zero values to 1 when casting. So `(bool)0xff` will evaluate to `00000001` in binary.

I have put together a short test program which applies arithmetic operations to boolean values and the result is the same for concrete evaluation using `gcc` to compile and run and with running the same program though `cbmc`. Therefore the behaviour appears to be correct from the perspective of a user of cbmc. Whether this is (A) undefined behaviour according to the C standard. (B) Implicitly casting booleans to integer types to perform the arithmetic or (C) applying the arithmetic operations directly to the boolean values, isn't a question I have an answer to yet.

In conclusion - whether `c_bool_typet` should derive from `bitvector_typet` as it does at the moment, or from `integer_bitvector_typet` or some other type is an open question to me. For what it is worth, I'd like to limit the scope of this PR to supporting the current data structures rather than a wider refactor of CBMC.

Comment (NlightNFotis) [reply src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: Agree with you re the scope of this.

Your answer satisfied my question.

**Review by NlightNFotis — APPROVED** [review]
